### PR TITLE
WaveQ loss first implementation

### PR DESCRIPTION
**Pull request by SKholkin**

### CODE REVIEW

**Comment by SKholkin** [comment]
@ljaljushkin

**Comment by SKholkin** [comment]
@ljaljushkin
comment
test

**Comment by ljaljushkin** [comment]
> @ljaljushkin
> comment
> test

please try to comment without using `@` symbol: just type 'test'

**Comment by SKholkin** [comment]
test

**Comment by ljaljushkin** [comment]
> test

yep, it works! I've received notifications with and without @-mentioning. So feel free to post some questions here. 
P.S. you may delete all 'test' comments to not distract from important messages

**Comment by SKholkin** [comment]
@ljaljushkin

**Comment by SKholkin** [comment]
@ljaljushkin
First implementation of Quantization Modules in WaveQ.
I tried to refactor WaveQ for being able to work with Quantization Modules, but i don't know how to test it and make sure it works (probably it doesn't) . Also hook has been added and so forward was refactored too. 
Main problem is that i don't how to make it work with NNCFNetwork and exactly how to get Quantize Modules   `def model_to_quantize_modules_convert` and write a test to proof usability.

**Comment by ljaljushkin** [comment]
> @ljaljushkin
> First implementation of Quantization Modules in WaveQ.
> I tried to refactor WaveQ for being able to work with Quantization Modules, but i don't know how to test it and make sure it works (probably it doesn't) . Also hook has been added and so forward was refactored too.
> Main problem is that i don't how to make it work with NNCFNetwork and exactly how to get Quantize Modules `def model_to_quantize_modules_convert` and write a test to proof usability.

you can create simple model and compress it.
see this test for the reference `test_quantization_configs__with_defaults`

File: nncf/quantization/waveq_loss.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+from nncf.compression_method_api import CompressionLoss
+from nncf.quantization.layers import QUANTIZATION_MODULES
+from nncf.utils import get_all_modules_by_type
+from nncf.nncf_network import NNCFNetwork
+
+
+class WaveQLoss(CompressionLoss):
+
+    def __init__(self, model: NNCFNetwork):
```
Comment (ljaljushkin) [inline nncf/quantization/waveq_loss.py]: don't need to pass the whole model. just use `QuantizationController.all_quantizations`

File: nncf/quantization/waveq_loss.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        self.model = model
+        # upper line doesn't work
+        # "cannot assign module before Module.__init__() call")
+        self.modules = model.children()
+        self.quantize_modules = self.model_to_quantize_modules_convert(self.model)
```
Comment (ljaljushkin) [inline nncf/quantization/waveq_loss.py]: please remove `model`, `modules` and `model_to_quantize_modules_convert` function as not needed

File: nncf/quantization/waveq_loss.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+            quantization_type = class_type.__name__
+            module_dict = get_all_modules_by_type(self.model, quantization_type)
+        return module_dict
+
+    def calc_quant_tensor_by_module(self):
+        self.hook_handlers = []
+        for module in self.quantize_modules:
+            hook_holder = LossHookHolder()
+            self.hook_handlers.append(torch.nn.Module.register_forward_hook(hook_holder.calc_hook()))
+        self.model.do_dummy_forward()
```
Comment (ljaljushkin) [inline nncf/quantization/waveq_loss.py]: no, forward should be called by user from sample. you just set up `hooks` ("observers") there.

File: nncf/quantization/waveq_loss.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        self.calc_quant_tensor_by_module()
+        for hooker in self.hook_handlers:
+            loss += WaveQLoss.waveq_loss_per_layer_sum(hooker.out_tensor)
+        return loss
+
+    def statistics(self):
+        # printable stats
+        pass
+
+    def get_loss_stats_tensor(self, ratio=1, quant=8):  # list
```
Comment (ljaljushkin) [inline nncf/quantization/waveq_loss.py]: this function is needed for tests only. let's move this code to test function.

File: tests/quantization/test_waveq_loss.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+    def set_rand_layers(self):
+        # try another distribution
+        for module in self.children():
+            module.weight.data = torch.randn(module.weight.data.shape)
+
+
+def test_graph_view(model):
+    waveq = WaveQLoss(model)
+    loss_tensors = waveq.get_loss_stats_tensor(ratio=16, quant=1)
```
Comment (ljaljushkin) [inline tests/quantization/test_waveq_loss.py]: you can check graph without model and `get_loss_stats_tensor` by using `waveq_loss_per_layer_tensor` for evenly generated data.

**Comment by SKholkin** [comment]
Module init from Quantization modules is done. Test was written, but  `compressed_model.do_dummy_forward()`  and some parts of forward environment doesn't work (probably because of wrong `compressed_model.do_dummy_forward()` usage). Also you can see some comments in the code and  all in all `forward` doesn't work.

File: tests/quantization/test_waveq_loss.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        layer_count += 1
+        plt.show()
+
+
+def test_model_to_quantize_converter():
+    model = BasicConvTestModel()
+    config = get_basic_quantization_config()
+    compressed_model, compression_ctrl = create_compressed_model_and_algo_for_test(model, config)
+    assert isinstance(compression_ctrl, QuantizationController)
+    loss_module = WaveQLoss(list(compression_ctrl.all_quantizations.values()))
```
Comment (ljaljushkin) [inline tests/quantization/test_waveq_loss.py]: In final version, you should create loss inside quantization controller and access it from sample using public API: 
`compression_loss = compression_ctrl.loss()`

File: tests/quantization/test_waveq_loss.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+    loss_module.set_up_hooks()
```
Comment (ljaljushkin) [inline tests/quantization/test_waveq_loss.py]: it should happen in constructor of WaveQLoss

File: tests/quantization/test_waveq_loss.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    something = compressed_model.do_dummy_forward()
+    # How to forward? and add weights to QM?
+    # SymmetricQuantizer.quantize(x) eqv SymmetricQuantizer.forward(x)
```
Comment (ljaljushkin) [inline tests/quantization/test_waveq_loss.py]: no, SymmetricQuantizer will be called automatically after creating compressed model

File: nncf/quantization/waveq_loss.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+from nncf.compression_method_api import CompressionLoss
+
+
+class WaveQLoss(CompressionLoss):
+
+    def __init__(self, quantize_modules):
+        self.quantize_modules = quantize_modules
+        self.hook_handlers = None
+
+    def set_up_hooks(self):
```
Comment (ljaljushkin) [inline nncf/quantization/waveq_loss.py:37]: After destruction of WaveQ loss all hooks still will be called. I think it makes sense to delete them in
`def __del__(self):`

File: tests/quantization/test_waveq_loss.py [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+        plt.show()
+
+
+def test_model_to_quantize_converter():
+    model = BasicConvTestModel(in_channels=1, out_channels=1, kernel_size=10, weight_init=0)
+    for child in model.children():
+        child.weight.data.normal_()
+    config = get_basic_asym_quantization_config(model_size=10)
+    # how to compress weights NOT INPUT
+    # cuz dummy_forward compresses only input
```
Comment (ljaljushkin) [inline tests/quantization/test_waveq_loss.py]: not really, it calls all Quantizers for both weight and activations. The difference with weights is that inputs for activations will be randomly generated, not actual images.

File: tests/quantization/test_waveq_loss.py [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+
+def test_waveq_loss_float_value_check():
+    layers = list()
+    layers.append(torch.nn.Conv2d(2, 2, 2, 2))
+    layers[0].weight.data = tensor_4d_float
+    layers.append(torch.nn.Conv2d(1, 1, 2, 2))
+    layers[1].weight.data = torch.tensor([[[[0.033], [0.0137]], [[0.0206], [0.0166]]]])
+    loss_module = WaveQLoss(convlayers=layers)
+    predicted_loss = 0.05612
+    assert math.isclose(loss_module.forward(), predicted_loss, rel_tol=0.005)
```
Comment (ljaljushkin) [inline tests/quantization/test_waveq_loss.py]: it's possible to use pytest.approx instead for better integration with pytest.

File: tests/quantization/test_waveq_loss.py [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+    writer = tb.SummaryWriter(log_dir=log_dir)
+
+    layer_count = 1
+    weights = []
+    for child in model.children():
+        weights.append(child.weight.data)
+
+    for weights_loss in zip(weights, loss_tensors):
+        weights_tensor = weights_loss[0]
+        loss_tensor = weights_loss[1]
```
Comment (ljaljushkin) [inline tests/quantization/test_waveq_loss.py]: you can use shorter syntax:
`for weights_tensor,  loss_tensor in zip(weights, loss_tensors):`

File: tests/quantization/test_waveq_loss.py [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+        loss_tensor = torch.flatten(loss_tensor)
+        weights_tensor = torch.flatten(weights_tensor)
+
+        zipped = sorted(zip(loss_tensor, weights_tensor), key=lambda tup: tup[1])
```
Comment (ljaljushkin) [inline tests/quantization/test_waveq_loss.py]: you can sort by using torch functionality in more natural way
`idx = torch.argsort(tensor)`

then access to sorted array by using given indexes
`sorted = tensor[idx]`

File: tests/quantization/test_waveq_loss.py [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+
+tensor_4d_float = torch.tensor([[[[0.002], [0.0098]],
+                                 [[0.0025], [0.0021]]],
+                                [[[0.0140], [0.098]],
+                                 [[0.033], [0.0166]]],
+                                [[[0.0206], [0.0284]],
+                                 [[0.04], [0.029]]],
+                                [[[0.0145], [0.0137]],
+                                 [[0.003], [0.0294]]]], dtype=torch.float32)
+log_dir = '/home/skholkin/projects/pycharm_storage/first_tasks/run'
```
Comment (ljaljushkin) [inline tests/quantization/test_waveq_loss.py]: you may consider using `tmp_path` fixture for tests. it creates a new temporary directory (like that `/tmp/pytest-of-nlyalyus/pytest-405`) for each test run, which is automatically deleted by pytorch after some number of runs.

example of usage:
```python
def test_export_lstm_cell(tmp_path):
test_path = str(tmp_path.joinpath('test.onnx'))
assert os.path.exists(test_path)
```

**Comment by SKholkin** [comment]
Добавил в QuantizationController в __init__() добавление WaveQ loss, и на моем примере с BasicConvTestModel оно работает, но в примере по классификации оно не работает. Причем вообще конструктоор для QuantizationController не вызывается и соотвественно WaveQ не добавляется.

File: nncf/quantization/waveq_loss.py [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+        self.hook_handlers = []
+        self.hook_collectors = []
+        for module in self.quantize_modules:
+            hook = LossHook(module)
+            self.hook_handlers.append(module.register_forward_hook(hook.calc_hook))
+            self.hook_collectors.append(hook)
+
+    def __call__(self, *args, **kwargs):
```
Comment (ljaljushkin) [inline nncf/quantization/waveq_loss.py]: please remove this. it's already in the base class

**Comment by SKholkin** [comment]
WaveQ Symmetric Quantization(unsigned) Update. Scale and bits(quant) params have been realized. Now WaveQ minimums match quantization levels (but there are some problems like by graphic WaveQ exactly matches quant levels but loss isn't exactly 0, for example while training mobilenet it's 100-200 times lower(when loss is calculated by quantized weights) but not exact 0). Also while training mobilenet CompLoss(calculated by input not quantized weights) doesn't lower and doesn't grow (although it's 6 times bigger).

**Comment by SKholkin** [comment]
Signed and Unsigned Symmetric Quantization Implementation finished and some graphic representation changes.

File: nncf/quantization/algo.py [hunk 16]
```diff
@@ -663,6 +664,10 @@ def __init__(self, target_model: NNCFNetwork,
 
         should_export_to_onnx_qdq = quantization_config.get("export_to_onnx_standard_ops",
                                                             False)
+        # TODO: config choice
+        if quantization_config['waveq']:
+            self._loss = WaveQLoss(list(self.all_quantizations.values()), ratio=1/600)
```
Comment (ljaljushkin) [inline nncf/quantization/algo.py]: In production it should look something like this:

```
"compression": {
"algorithm": "quantization",
"params": {
      "loss": "waveq",
      "ratio": 1/600,
}}
```

**Comment by SKholkin** [comment]
Some WaveQ module fixes connected with type of quantization and tool update so now you can visualize info from logs.

File: nncf/quantization/algo.py [hunk 17]
```diff
@@ -663,6 +664,11 @@ def __init__(self, target_model: NNCFNetwork,
 
         should_export_to_onnx_qdq = quantization_config.get("export_to_onnx_standard_ops",
                                                             False)
+        try:
```
Comment (ljaljushkin) [inline nncf/quantization/algo.py]: no need to use try, except there. just use `get` method (https://www.tutorialspoint.com/python/dictionary_get.htm) with some default value, like None or ''.

File: tests/quantization/test_algo_quantization.py [hunk 18]
```diff
@@ -58,7 +58,7 @@ def get_basic_quantization_config(model_size=4):
                 "algorithm": "quantization",
                 "initializer": {
                     "range": {
-                        "num_init_steps": 0
+                        "num_init_steps": 1
```
Comment (ljaljushkin) [inline tests/quantization/test_algo_quantization.py]: please don't modify configs for other tests

File: tests/quantization/test_waveq_loss.py [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+import torch
+import math
+from torch.utils import tensorboard as tb
+import matplotlib.pyplot as plt
+from tests.test_helpers import BasicConvTestModel, create_compressed_model_and_algo_for_test
+from tests.quantization.test_algo_quantization import get_basic_quantization_config, \
+    get_basic_asym_quantization_config, OnesDatasetMock
+from nncf.initialization import register_default_init_args
+
+kernel_size = 10
```
Comment (ljaljushkin) [inline tests/quantization/test_waveq_loss.py]: avoid using of global variables and definitions - it's [evil](https://stackoverflow.com/questions/19158339/why-are-global-variables-evil) `pytest.fixture` is more suitable. there're many examples in repo.

File: nncf/quantization/waveq_loss.py [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+    def waveq_loss_per_hook_sum(hook: LossHook, ratio=1):
+        input_low, input_range, levels = get_quant_module_params(hook.quant_module)
+        out = WaveQLoss.waveq_loss_for_tensor(hook.input_tensor, ratio,
+                                              levels=levels, input_low=input_low, input_range=input_range)
+        return torch.sum(out)
+
+
+def get_quant_module_params(quant_module: BaseQuantizer):
+    try:
+        quant_module.signed
```
Comment (ljaljushkin) [inline nncf/quantization/waveq_loss.py]: what if exception happens later? it's wrong logic. quick & dirty fix is to use `isinstance()` method. 
But I would add to `BaseQuantizer` `get_quantization_ranges` method and return final input_low and input_range. 

Please pay attention to some modifications of parameters:
for symmetric
```
scale_safe = abs(scale) + eps
```
and assymetric
```
    input_range_safe = abs(input_range) + eps
    input_low_tuned, input_range_tuned = TuneRange.apply(input_low, input_range_safe, levels)
```

Without these modifications WaveQ loss can be inaccurate.

**Comment by SKholkin** [comment]
BaseQuantizer, SymmetricQuantizer, AsymmetricQuantizer modules have been refactored. Now there are `calculate_all_ranges` function that calculates levels and inputs, but inputs that are calculated in this function are not inputs to be set into class variables so they constructed for usage outside the module. Also there are `updated_ranges` that is used for calculating especially updated  inputs for usage outside of the class too.

**Comment by ljaljushkin** [comment]
> BaseQuantizer, SymmetricQuantizer, AsymmetricQuantizer modules have been refactored. Now there are `calculate_all_ranges` function that calculates levels and inputs, but inputs that are calculated in this function are not inputs to be set into class variables so they constructed for usage outside the module. Also there are `updated_ranges` that is used for calculating especially updated inputs for usage outside of the class too.

please run old tests and make sure everything is not broken.

File: nncf/quantization/layers.py [hunk 21]
```diff
@@ -197,33 +200,41 @@ def __init__(self, config):
...
+            level_high = 2 ** self.num_bits - 1
             level_low = 0
-        levels = 2 ** num_bits
-        if is_weights:
-            levels -= 1
-        return level_high, level_low, levels
+
+        scale = abs(self.scale) + self.eps
+        input_low, input_range = self.updated_ranges()
+        return input_low, input_range, level_high, level_low, levels
```
Comment (ljaljushkin) [inline nncf/quantization/layers.py]: it's not a good practice to return too many arguments. try to avoid it

File: nncf/quantization/layers.py [hunk 22]
```diff
@@ -234,7 +245,8 @@ def signed(self, signed: bool):
         self.signed_tensor.fill_(signed)
 
     def quantize(self, x):
-        return symmetric_quantize(x, self.levels, self.level_low, self.level_high, self.scale, self.eps)
+        input_low, input_range = self.updated_ranges()
+        return symmetric_quantize(x, input_low, input_range, self.levels, self.level_low, self.level_high)
```
Comment (ljaljushkin) [inline nncf/quantization/layers.py]: do you really need 2 last arguments there? `self.level_low, self.level_high`
